### ISSUE TITLE
FTP support

### ISSUE THREAD

**Issue by martinschorb** [mobie/mobie-viewer-fiji#110]
Hi,

now that I am uploading data to EMPIAR, I found that all data there is available via FTP, hooray!
example:
`ftp://ftp.ebi.ac.uk/empiar/world_availability/10050/data/`

Assuming I upload n5 "files" (need to check with the EBI guys, whether that's OK) it should be rather straight forward to make MoBIE accept an EMPIAR ID as project input, correct?

It would then need a check of the data type (either by looking for a key file, or through the EMPIAR API) whether the entry is in the proper format.

Am I missing something, or is it really that easy, and the biggest issue is convincing the EBI people of the n5 format and the different kind of traffic pattern when only accessing small chunks of data instead of full volumes?

**Comment by constantinpape** [comment]
The problem is that there is no ftp implementation for n5 yet.

Building one in python would be rather straightforward, using [ftlib](https://docs.python.org/3/library/ftplib.html) and then decoding chunks either with [zarr](https://github.com/zarr-developers/zarr-python) or [z5py](https://github.com/constantinpape/z5).
(Maybe even easier via [pyfilesystem](https://github.com/PyFilesystem/pyfilesystem)).

I am not sure how much work this would be in java though, which is what we would need for the mobie-viewer.

Related to this, there is a project pending for connecting EBI-Services and MoBIE, I am not quite sure what the current status of this is, maybe @tischi knows more.

**Comment by tischi** [comment]
> Related to this, there is a project pending for connecting EBI-Services and MoBIE, I am not quite sure what the current status of this is, maybe @tischi knows more.

We submitted a project proposal but I don't know when the decision for this to be granted will be taken.
However, the ftp connection is interesting. I have no idea how performant it is.

If the performance is ok, developing an n5ftp java library would be within scope, as the implementations are not much code, see e.g.: https://github.com/saalfeldlab/n5-aws-s3/blob/master/src/main/java/org/janelia/saalfeldlab/n5/s3/N5AmazonS3Reader.java

@martinschorb could you tell me how the ftp call to one chunk of the data would look like?

**Comment by martinschorb** [comment]
Hi,

I have tested the download performance for the dataset linked above and could reach ~10 MB/s, so very comparable what EMBL's s3 delivers.

I have not yet uploaded a test n5 dataset since uploading only works with a proper deposition. I am in touch with the EBI people (will CC you) to see if we can give them a test dataset to upload and try the connection.

You basically upload the data in entire directories and can control how they are organized yourself, so I think we could upload the n5 structure "as is".

Basically the link to the data would then look like:

ftp://ftp.ebi.ac.uk/empiar/world_availability/$EMPIAR_ID/data/$dataset_name/images/... and then the n5 structure as in aws

**Comment by martinschorb** [comment]
Let's see what they respond in terms of file format(s). I don't kow if they are willing to allow n5 (quoting the n5 github: "Please do not take this project too seriously, we will see where it will get us and report back when more data is available.") or rather wait until there is some specification for the OME format.

**Comment by tischi** [comment]
> or rather wait until there is some specification for the OME format.

The project proposal that I have submitted indeed would probably end up using ome-zarr.

**Comment by constantinpape** [comment]
> The project proposal that I have submitted indeed would probably end up using ome-zarr.

Yes, I think we should switch to zarr at some point in any case, but the whole ome-zarr / zarr spec v3 process is still developing:
https://github.com/zarr-developers/zarr-specs/issues

**Comment by constantinpape** [comment]
I don't think this is something we want to go for, as we would need to reimplement an `N5FTPReader`. Instead, we want to extend support for ome.zarr and then push EMPIAR/IDR/bioimagearxiv to accept it.
